### ISSUE TITLE
Change Request: How foo/**/* is handled in "files" key

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#15743]
### ESLint version

HEAD

### What problem do you want to solve?

The current `ESLint` implementation with overrides specifies that a pattern of `foo/**/*` will match files only with a `.js` pattern when a directory is passed on the command line:

https://github.com/eslint/eslint/blob/8bb527660ffddc122204fb5bf59076687aaa78b8/tests/lib/eslint/eslint.js#L6073-L6093

As best I can tell, this happens because the preconfigured file extensions only include `.js` in the current version, so `*` is taken to mean `*.js`.

With flat config, `foo/**/*` currently matches all files regardless of extension when a directory is passed on the command line.

### What do you think is the correct solution?

I'm not 100% sure. There are options:

1. We could mimic the current behavior of ESLint, which is that any pattern ending with `*` will automatically be appended with configured file extensions. In this case, that would means anything in the flat config in the format `**/*.ext` or anything passed with `--ext`.
2. Keep the current flat config behavior, which is that `*` will match any file regardless of extension.

I favor option 2 here, as I think this is consistent with how patterns are used inside of flat config. Even though this deviates from the current behavior of `overrides`, this seems like the best choice for internal consistency.

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by mdjermanovic** [comment]
> 1. We could mimic the current behavior of ESLint, which is that any pattern ending with `*` will automatically be appended with configured file extensions. In this case, that would means anything in the flat config in the format `**/*.ext` or anything passed with `--ext`.

Patterns ending with `*` don't apply when determining whether a file should be linted, but they do apply when calculating the config for that file. Patterns that don't end with `*` always apply when determining whether a file should be linted, regardless of their format (it doesn't have to be `*.ext` or `**/*.ext`). For example:

```js
//--------- mydir/foo.txt ----------

void undef_var
```

```js
//-------- .eslintrc.js ----------

module.exports = {
    overrides: [
        {
            files: ["mydir/foo.txt"],
            rules: {
                "no-undef": 2
            },
        },
        {
            files: ["mydir/**/*"],
            rules: {
                "no-void": 2
            },
        }
    ],
};
```

```
npx eslint .
```

```
D:\tmp\tmp\mydir\foo.txt
  3:1  error  Expected 'undefined' and instead saw 'void'  no-void
  3:6  error  'undef_var' is not defined                   no-undef

✖ 2 problems (2 errors, 0 warnings)
```

**Comment by mdjermanovic** [comment]
> 1. We could mimic the current behavior of ESLint, which is that any pattern ending with `*` will automatically be appended with configured file extensions. In this case, that would means anything in the flat config in the format `**/*.ext` or anything passed with `--ext`.
> 2. Keep the current flat config behavior, which is that `*` will match any file regardless of extension.
> 
> I favor option 2 here, as I think this is consistent with how patterns are used inside of flat config. Even though this deviates from the current behavior of `overrides`, this seems like the best choice for internal consistency.

I also favor option 2. It's much easier to implement and maintain, and I believe it would be much easier for users to understand how this works.

A downside is that users would have to add extensions (`foo/**/*` -> `foo/**/*.js`) to each glob that matches files that shouldn't be linted.

**Comment by ljharb** [comment]
Option 2 seems like it would be a ton of false positives. I have tons of non-lintable files that would get included, and being forced to provide the extensions in the overrides would scatter that information in multiple places - i'd prefer the list of lintable extensions be defined in as few places as possible.

**Comment by nzakas** [comment]
> Patterns ending with * don't apply when determining whether a file should be linted, but they do apply when calculating the config for that file.

@mdjermanovic ah, I see! (My mind is being warped by spending too much time in these tests.) Since that's the case, I can easily filter out patterns ending with `*` when coming up with glob patterns to find files. To @ljharb's point, the downside of option 2 is that it somewhat breaks backwards compatibility with `overrides` going forward.

The change to match the current behavior turned out to be very simple based on how I'm doing the file search:

```js
if (filePattern.endsWith("/*")) {
    return false;
}
```

Based on the ease of implementation and backwards-compatibility, it seems like Option 1 is probably the way to go at this point.
